# Archive of software used for Hansen et al. (2025). A complete diploid human genome benchmark for personalized genomics
### Authors
Nancy F. Hansen1, Nathan Dwarshuis2, Hyun Joo Ji3,4, Arang Rhie1, Hailey Loucks5, Glennis A. Logsdon6, Mitchell R. Vollger7, Jessica M. Storer8, Eleni Adam9, Nicolas Altemose10,11, Dmitry Antipov1, Mobin Asri5, Sofia Barreira12, Stephanie C. Bohaczuk7, Andrey V. Bzikadze13, Sara A. Carioscia14, Andrew Carroll15, Kuan-Hao Chao3,4, Yanan Chu16, Arun Das3, Peter Ebert17,18, Adam English19, Mark Fleharty20, Laura E. Fleming20, Giulio Formenti21, Andrea Guarracino22, Gabrielle A. Hartley8, Katharine Jenike3,23, Jenna Kalleberg24, Yu Kang16, Juhyun Kim1, Robert King25, Josipa Lipovac26, Mira Mastoras5, Matthew W. Mitchell27, Shloka Negi5, Nathan D. Olson2, Keisuke K. Oshima6, Luis F. Paulin19, Brandon D. Pickett1, David Porubsky28,29, Jane Ranchalis7, Desh Ranjan9, Mikko Rautiainen30, Harold Riethman31, Robert D. Schnabel24, Fritz J. Sedlazeck19,32,33, Kishwar Shafin15, Mile Sikic26,34, Steven J. Solar1, Alexander P. Sweeten1,3, Winston Timp35, Justin Wagner2, DongAhn Yoo28, Ying Zhou36, Erik Garrison22, Evan E. Eichler28,37, Michael C. Schatz3, Andrew B. Stergachis7,28,38, Rachel J. O'Neill8,39,40, Karen H. Miga5, Steven L. Salzberg3,4,35,41, Sergey Koren1, Justin M. Zook2,\*, Adam M. Phillippy1,\*

### Affiliations
1 Genome Informatics Section, Center for Genomics and Data Science Research, National Human Genome Research Institute, National Institutes of Health, Bethesda, MD, USA
2 Material Measurement Laboratory, National Institute of Standards and Technology, Gaithersburg, MD, USA
3 Department of Computer Science, Johns Hopkins University, Baltimore, MD, USA
4 Center for Computational Biology, Johns Hopkins University, Baltimore, MD, USA
5 UC Santa Cruz Genomics Institute, University of California, Santa Cruz, CA, USA
6 Department of Genetics, Epigenetics Institute, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA, USA
7 Division of Medical Genetics, University of Washington School of Medicine, Seattle, WA, USA
8 Institute for Systems Genomics, University of Connecticut, Storrs CT, USA
9 Department of Computer Science, Old Dominion University, Norfolk, VA, USA
10Department of Genetics, School of Medicine, Stanford University, Palo Alto, CA, USA 
11Chan Zuckerberg Biohub, San Francisco, San Francisco, CA USA
12Computational Genomics Unit, Center for Genomics and Data Science Research, National Human Genome Research Institute, National Institutes of Health, Bethesda, MD, USA
13 Graduate Program in Bioinformatics and Systems Biology, University of California, San Diego, La Jolla, CA, USA
14 Department of Biology, Johns Hopkins University, Baltimore, MD, USA
15 Google LLC, Mountain View, CA, USA
16 Beijing Institute of Genomics, Chinese Academy of Sciences, Beijing, 100101, China
17 Core Unit Bioinformatics, Medical Faculty and University Hospital Düsseldorf, Heinrich Heine University, Düsseldorf, Germany
18 Center for Digital Medicine, Heinrich Heine University Düsseldorf, Germany
19 Human Genome Sequencing Center, Baylor College of Medicine, Houston, TX, USA
20 Broad Institute, Broad Clinical Labs, Burlington, MA, USA
21 The Vertebrate Genome Laboratory, The Rockefeller University, New York, USA
22 Department of Genetics, Genomics and Informatics, University of Tennessee Health Science Center, Memphis, TN, USA
23 Department of Plant Sciences, University of Cambridge, Cambridge, UK, CB2 3EA
24 Division of Animal Sciences, University of Missouri, Columbia, MO, USA
25 Oxford Nanopore Technologies, Oxford, UK
26 Laboratory for Bioinformatics and Computational Biology, Faculty of Electrical Engineering and Computing, University of Zagreb, Zagreb, Croatia
27 Coriell Institute for Medical Research, Camden, NJ, USA
28 Department of Genome Sciences, University of Washington School of Medicine, Seattle, WA, USA
29 Genome Biology Unit, European Molecular Biology Laboratory (EMBL), Heidelberg, 69117, Germany
30 Institute for Molecular Medicine Finland, Helsinki Institute of Life Science, University of Helsinki, Helsinki, Finland
31School of Medical Diagnostic & Translational Sciences, Old Dominion University, Norfolk VA, USA
32 Department of Molecular and Human Genetics, Baylor College of Medicine, TX, USA
33 Department of Computer Science, Rice University, Houston, TX, USA
34 Genome Institute of Singapore, A\*STAR
35 Department of Biomedical Engineering, Johns Hopkins University, Baltimore, MD, USA
36 Department of Data Science, Dana-Farber Cancer Institute, Boston, MA, USA
37 Howard Hughes Medical Institute, University of Washington, Seattle, WA, USA
38 Brotman Baty Institute for Precision Medicine, Seattle, WA, USA
39 Dept of Molecular and Cell Biology, University of Connecticut, Storrs CT, USA
40  Dept of Genome Sciences, UConn Health, Farmington, CT, USA
41 Department of Biostatistics, Johns Hopkins University, Baltimore, MD, USA

* Corresponding authors: Justin M. Zook <justin.zook@nist.gov>, Adam M. Phillippy <adam.phillippy@nih.gov>

### Repository contents
This repository contains code, analysis details, and descriptions and copies of github repositories used in the analyses for the manuscript "A complete diploid human genome benchmark for personalized genomics" by Hansen et al. A release of the repository will be archived and cited when the manuscript is submitted.
### Acknowledgements
This work was supported, in part, by the Intramural Research Program of the US National Human Genome Research Institute, National Institutes of Health (NFH, SK, AR, DA, JK, BDP, SJS, APS, and AMP). The contributions of NIH authors are considered Works of the United States Government. The findings and conclusions presented in this paper are those of the authors and do not necessarily reflect the views of the NIH or the U.S. Department of Health and Human Services. This work is also supported, in part, by NIST Intramural Funding (ND, NDO, JW, and JMZ). Certain commercial equipment, instruments, or materials are identified to specify adequately experimental conditions or reported results. Such identification does not imply recommendation or endorsement by the National Institute of Standards and Technology, nor does it imply that the equipment, instruments, or materials identified are necessarily the best available for the purpose. Several authors are supported by NIH grants R01-HG006677 (to HJJ), 5R01HG010169 (to EEE), R00GM147352 (to GAL), 2T32GM007454, 1K99GM155552, and  1U01HG013744 (to MRV), R01-HG006677 and R35-GM156470 (to K-HC), U24HG010263 (to AD), UM1DA058229 (to LFP), R01-HG006677 and R35-GM130151 (to SLZ), 1UG3NS132105 and 1U01HG011758 (to FJS), U24HG010263 (to MCS), 1DP5OD029630, UM1DA058220, and 1U01HG013744 (to ABS) as well as USDA NIFA 2023-67015-39261 (to JK and RDS) and HATCH MO HAAS0001 (to RDS). NA is a Chan Zuckerberg Biohub Investigator and is supported by an HHMI Hanna H. Gray Fellowship and a Pew Biomedical Scholar Award. RJO is funded by the John and Donna Krinicki Endowment and by the Colossal Foundation. EEE is an investigator of the Howard Hughes Medical Institute. This work utilized the computational resources of the NIH HPC Biowulf cluster (http://hpc.nih.gov).

